ebfc/ebfc.1
.TH EBFC 1 "May 2021" "ELF kickers 3.2"
.LO 1
.SH NAME
ebfc \- ELF Brainfuck compiler
.SH SYNOPSIS
.B ebfc
[OPTIONS]
.I SRCFILE
.SH DESCRIPTION
.B ebfc
is a compiler for the Brainfuck programming language, creating 64-bit
ELF files targeted for the Intel x86-64 architecture.
.P
.B ebfc
can create standalone executables, shared libraries, or object files.
Object files can themselves be targeted for a standalone executable, a
shared library, or a module in a larger program.
.SH COMPILATION TARGET OPTIONS
The following options control what type of source file is created. It
is assumed here that the name of the source file is
.BR foo.b .
.TP
.B \-x
Compile the source code into a standalone executable file, named
.BR foo .
.TP
.B \-l
Compile the source code into a shared library file, named
.BR libfoo.so .
The program will be exported as a function named
.BR foo ().
.TP
.B \-c
Compile the source code into a function, named
.BR foo (),
in an object file, named
.BR foo.o .
The object file will be targeted for a module in an executable.
.TP
.B \-xc
Compile the source code into an object file, named
.BR foo.o ,
that can then be linked as a standalone executable.
.TP
.B \-lc
Compile the source code into a function, named
.BR foo (),
in an object file, named
.BR foo.o ,
that can then be linked as a shared library.
.P
If the
.I SRCFILE
argument lacks a
.B .b
or
.B .bf
suffix, then the entire filename will be used when creating the name
of the target file and function. (In the case of
.IR -x ,
the name of the target file will be
.B a.out
instead.)
.SH OTHER OPTIONS
.TP
\fB\-a\fR, \fB\-\-arg\fR
Compile the function to take an argument, namely a pointer to the byte
array. By default,
.B ebfc
will compile its code to a function that takes no arguments, which
will use an internal byte array. When this option is used, there is no
difference in the object file created by
.I -c
and
.IR -lc .
(This option is not applicable to a standalone executable.)
.TP
\fB\-f\fR, \fB\-\-function=\fIFUNCTION\fR
Use
.I FUNCTION
as the name of the function to contain the compiled program. If this
argument is omitted, then the function's name will be taken from the
source filename, as described in the previous section.
.TP
\fB\-i\fR, \fB\-\-input=\fIFILE\fR
Use
.I FILE
as the name of the source file to place in the target file. This
option does not supersede the
.I SRCFILE
argument; it merely changes what name is stored in the object file.
(This option is not meaningful if the target is not an object file, or
if the
.I \-\-strip
option is used.)
.TP
\fB\-o\fR, \fB\-\-output=\fIFILE\fR
Use
.I FILE
as the target filename. If this option is omitted, the output filename
will be generated from the source filename, as described in the
previous section.
.TP
\fB\-s\fR, \fB\-\-strip\fR
Suppress inclusion of unnecessary data in the target file. By default,
.B ebfc
includes a .comment section, and includes a symbol in the symbol table
for the source filename. These items will be removed from the output
when this option is used. Additionally, if the output is a standalone
executable, this option will suppress inclusion of the section header
table.
.TP
\fB\-z\fR, \fB\-\-compressed\fR
Read the source file in compressed Brainfuck format (see below).
.TP
.B \-\-help
Display help and exit.
.TP
.B \-\-version
Display version number and exit.
.SH LINKING
When calling a compiled Brainfuck program from within a C program, the
C prototype for the function should have the form:
.P
    extern void foo(void);
.P
or, if the
.I \-\-arg
option is specified:
.P
    extern void foo(char *buffer);
.P
In the latter case, the buffer pointed to by the function parameter
will be used as the initial state of the Brainfuck buffer when the
function is invoked, and upon return will contain the final state of
the buffer.
.SH THE BRAINFUCK PROGRAMMING LANGUAGE
A Brainfuck program has an implicit byte pointer, called "the
pointer", which is free to move around within an array of bytes,
initially all set to zero. The pointer itself is initialized to point
to the beginning of this array. (The size of the array is not
constrained by the language, but is typically 30000 bytes or more.
.B ebfc
programs are given an array of 32768 bytes.)
.P
The Brainfuck programming language consists of eight commands, each of
which is represented as a single character.
.TP 4
.PD 0
.B >
Increment the pointer.
.TP
.B <
Decrement the pointer.
.TP
.B +
Increment the byte at the pointer.
.TP
.B \-
Decrement the byte at the pointer.
.TP
.B .
Output the byte at the pointer.
.TP
.B ,
Input a byte and store it in the byte at the pointer.
.TP
.B [
Jump to the matching
.B ]
if the byte at the pointer is zero.
.TP
.B ]
Jump to the matching
.BR [ .
.PD 1
.P
Any other characters in the source code are treated as comments or
whitespace, and ignored.
.P
The semantics of the Brainfuck commands can also be succinctly
expressed in terms of C, as follows (assuming that
.I p
has been previously defined as a
.IR char* ):
.TP 4
.PD 0
.B >
becomes\ \ ++p;
.TP
.B <
becomes\ \ \-\-p;
.TP
.B +
becomes\ \ ++*p;
.TP
.B \-
becomes\ \ \-\-*p;
.TP
.B .
becomes\ \ putchar(*p);
.TP
.B ,
becomes\ \ *p = getchar();
.TP
.B [
becomes\ \ while (*p) {
.TP
.B ]
becomes\ \ }
.PD 1
.P
As with C, the generated program's behavior is undefined if the
pointer is moved outside of the byte array.
.SH COMPRESSED BRAINFUCK
There is a compressed format for storing and transmitting Brainfuck
programs, which
.B ebfc
can read natively by using the
.I \-\-compressed
option.
.P
In compressed Brainfuck, the eight commands are encoded in three bits
as follows:
.TP 4
.PD 0
.B +
000
.TP
.B \-
001
.TP
.B <
010
.TP
.B >
011
.TP
.B [
100
.TP
.B ]
101
.TP
.B ,
110
.TP
.B .
111
.PD 1
.P
Each byte in a compressed Brainfuck file contains one or more
commands. The top two bits select between one of four possible
readings of the lower six bits, as follows:
.TP 24
.PD 0
Encoding\ \ \ \ Bits
\ Translation
.TP
singleton\ \ \ 00 abc abc
\ abc
.TP
pair\ \ \ \ \ \ \ \ 00 abc def
\ abc followed by def
.TP
triplet\ \ \ \ \ 10 ab cd ef
\ 0ab then 0cd then 0ef
.TP
repetition\ \ 01 abc def
\ def repeated 2 + abc times (2-9)
.TP
repetition\ \ 11 abcd ef
\ 0ef repeated 2 + abcd times (2-17)
.PD 1
.P
.SH ERRORS
The compiler will issue an error message, and the compilation will
fail, if the program contains unbalanced bracket commands, or if the
level of nested brackets exceeds the compiler's maximum capacity
(which is arbitrarily set at 256).
.SH COPYRIGHT
Copyright \(co 1999, 2001, 2021 Brian Raiter
.IR <breadbox@muppetlabs.com> .
.P
License GPLv2+: GNU GPL version 2 or later. This is free software: you
are free to change and redistribute it. There is NO WARRANTY, to the
extent permitted by law.


Changelog
* May 2021: version 3.2

	It's been a decade since 3.0 was released, wherein I noted my
	intention to update ebfc to support 64-bit ELF in the near future.
	ebfc is the only program in ELF Kickers that creates an ELF file
	from scratch, rather than working with an existing ELF file, and
	this made it trickier to support both 32-bit and a 64-bit targets
	in one program. However, at this point in time, it is wholly
	justifiable to simply drop support for 32-bit targets entirely
	(particularly with a program like ebfc, which is more intended to
	be a learning tool than an actual utility). So that's what I did.

	Along the way, I reworked ebfc's code to make it a little clearer
	and more consistent, and one or two features were added that
	didn't make it into the original 2.0 version. And, as I did ten
	years ago with the rest of ELF Kickers, the code has been cleaned
	up to be warning-free with current compilers.

	Finally, this release contains a handful of fixes to elfls,
	elftoc, and infect -- mainly for crashes on malformed input files.

* November 2019: version 3.1a

	Another very minor update. This release includes a couple of bug
	fixes, and the --force-hex option to elftoc.

	Thanks to Aapo Rantalainen, Agostino Sarubbo, and Lars Lindqvist
	for bug reports and patches.
	
* November 2016: version 3.1

	It's been five years since the last version, and a handful of bug
	fixes (and one feature) have accumulated over that time. A quick
	rundown:

	- objres: Added --reference option to improve non-x86 support.
	- elftoc: Make signed chars explicit where needed.
	- elftoc: Bug fix in ELF header version check.
	- elfls: Bug fix in handling Elf32 on a 64-bit system.
	- Various changes to the Makefiles.
	- Fixes for some typos in the documentation and comments.

	Thanks to German Zyuzin, Lars Lindqvist, Douglas Mencken, Aapo
	Rantalainen, and Steven Honeyman (in no particular order) for bug
	reports and patches.

* November 2011: version 3.0a

	This is a very minor update.

	Sofian Brabez supplied me with information to help ELF Kickers
	compile under FreeBSD. I've gratefully incorporated his changes.

	Raphaël Huck reported that sstrip's alteration to the e_shentsize
	field is flagged as invalid by file(1). This has been fixed.

	infect now attempts to reset the infected file's mtime. Not
	important, but why not.
	
* May 2011: version 3.0

	After a decade hiatus, I finally took the plunge and updated this
	package. With some effort I've managed to get a new version out in
	time for the ten-year anniversary of the last update.

	The most important change: All the programs (except ebfc) have
	been updated to deal equally well with 32-bit and 64-bit ELF
	files. This has been a long time coming (though it had to wait
	until I had a 64-bit machine of my own to test on).

	A very simple library, elfrw, has been added which hides many of
	the differences between 32-bit and 64-bit ELF files, allowing the
	smaller programs to support both types of files without
	significant code changes.

	The program objres has been added to the collection.

	The sample program infect has been added to the collection.

	All of the old programs have been cleaned up to avoid the new gcc
	warnings, use getopt_long, and in general brought more in line
	with my current programming style.

	elftoc in particular has had several sections rewritten. The most
	important change is that the build process reads the local copy of
	elf.h to find out which ELF macros are available on the current
	installation, instead of relying on a hard-coded list included in
	the distribution. elftoc also required lots of nearly-duplicate
	code in order to support both 32-bit and 64-bit ELF files.

	Man pages have been added for the utilities that didn't have them,
	for the sake of consistency if nothing else.

	The tiny programs have been removed from ELF kickers. Not because
	I don't love them any less, but they tend to suffer more quickly
	than the rest of the code from changes to Linux. In the last ten
	years I've had to modify nearly all of them at least once to keep
	them working with current kernels.

	(ebfc is the only utility that has not been updated for 64-bit. I
	intend to remedy this in the near future.)
	
* May 2001: version 2.0a

	Two small bug fixes, reported to me by people who downloaded
	version 2.0, the details of which are lost to history.
	
* April 2001: version 2.0

	I began spelunking in the system header files <elf.h> and
	<linux/elf.h> again, and discovered a large number of new
	definitions. I realized that the ELF standard could be going
	through major changes, and that having a version number is really
	necessary.

	elftoc updated to understand some of these new additions, and
	elsewhere restructured so as to make it easier to update it again
	once I understand the rest of them.

	Fixes for elftoc bugs that I noticed during the rewrite.

	elfls updated similarly to elftoc.

	Minor rewrite of sstrip to make the code more airtight and
	aesthetically pleasing.
	
* March 2001

	rebind added to ELF Kickers.

	date and snake added to the tiny collection.

	cat removed from the tiny collection (too uninteresting).

	Cleaned up all tiny programs and made them more stylistically
	consistent.

	Minor bug fix in elftoc.

	Documentation cleanups.

* February 2001

	Five of the tiny programs did not work under 2.2.17 and later 2.2
	kernels due to a change in how Linux checks the ELF header.
	(Thanks to Francois-Rene Rideau, who pointed this out to me. Oddly
	enough, this new restriction does not seem to have made it into
	the 2.4.0 kernel. I am unsure if the change is confined to the 2.2
	versions, or if it will resurface in a later kernel.) I succeeded
	in fixing two of the programs without sacrificing compatibility
	with 2.0 kernels or increasing the size. For the other three, I
	added new versions specifically for 2.2.17 (et al).

* January 2001

	keepalive added to the tiny collection.


ebfc/ebfc.c
/* ebfc.c: The central module.
 *
 * Copyright (C) 1999,2001,2021 by Brian Raiter, under the GNU General
 * Public License, version 2. No warranty. See COPYING for details.
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <stdbool.h>
#include <string.h>
#include <ctype.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <getopt.h>
#include <elf.h>
#include "elfparts.h"
#include "ebfc.h"

/* The online help text.
 */
static char const      *yowzitch =
    "Usage: ebfc [OPTIONS] FILENAME\n"
    "Compiles Brainfuck source code to an ELF target.\n\n"
    "  -x                    Generate a standalone executable file.\n"
    "  -l                    Generate a shared library.\n"
    "  -c                    Generate the object file for a function.\n"
    "  -xc                   Generate the object file for an executable.\n"
    "  -lc                   Generate the object file for a shared library.\n"
    "  -a, --arg             Modify the function to take a buffer argument.\n"
    "  -z, --compressed      Read the input file as compressed BF source.\n"
    "  -o, --output=FILE     Write the output to FILE.\n"
    "  -f, --function=NAME   Use NAME as the library's exported function.\n"
    "  -i, --input=NAME      Record the source filename as NAME.\n"
    "  -s, --strip           Omit non-required data from the output file.\n"
    "      --help            Display this help and exit.\n"
    "      --version         Display version information and exit.\n";

/* The version text.
 */
static char const *vourzhon =
    "ebfc, version 2.0\n"
    "Copyright (C) 1999,2021 by Brian Raiter <breadbox@muppetlabs.com>\n"
    "License GPLv2+: GNU GPL version 2 or later.\n"
    "This is free software; you are free to change and redistribute it.\n"
    "There is NO WARRANTY, to the extent permitted by law.\n";

/* The full collection of elfparts used by this program. This list
 * must match up with the enum that appears in the header file.
 */
static elfpart const *parttable[P_COUNT] = {
    [P_EHDR]     = &part_ehdr,
    [P_PHDRTAB]  = &part_phdrtab,
    [P_HASH]     = &part_hash,
    [P_DYNSYM]   = &part_dynsym,
    [P_DYNSTR]   = &part_dynstr,
    [P_TEXT]     = &part_text,
    [P_REL]      = &part_rel,
    [P_GOT]      = &part_got,
    [P_DYNAMIC]  = &part_dynamic,
    [P_DATA]     = &part_data,
    [P_COMMENT]  = &part_progbits,
    [P_SHSTRTAB] = &part_shstrtab,
    [P_SYMTAB]   = &part_symtab,
    [P_STRTAB]   = &part_strtab,
    [P_SHDRTAB]  = &part_shdrtab,
};

/* Lists indicating which elfparts are included in which file types.
 */
static int partlists[][P_COUNT] = {
    [ET_EXEC] = { P_PHDRTAB, P_TEXT, P_DATA, P_COMMENT,
                  P_SHSTRTAB, P_SHDRTAB },
    [ET_REL] =  { P_TEXT, P_REL, P_DATA, P_COMMENT,
                  P_SHSTRTAB, P_SYMTAB, P_STRTAB, P_SHDRTAB },
    [ET_DYN] =  { P_PHDRTAB, P_HASH, P_DYNSYM, P_DYNSTR, P_TEXT,
                  P_GOT, P_DYNAMIC, P_DATA, P_COMMENT, P_SHSTRTAB, P_SHDRTAB }
};

/* The name of this program, taken from argv[0]. Used for error messages.
 */
static char const *thisprog;

/* The name of the currently open file. Used for error messages.
 */
static char const *thefilename;

/* The filename to insert in the object file as the source filename.
 */
static char *srcfilename = NULL;

/* The filename to receive the object code.
 */
static char *outfilename = NULL;

/* The name of the function to put the Brainfuck program under.
 */
static char *functionname = NULL;

/* Whether or not to add extra items to the object file.
 */
static bool addextras = true;

/*
 * General-purpose functions.
 */

/* Output a formatted error message on stderr, prefixed by the input
 * filename. If fmt is NULL, then the error message supplied by
 * perror() is used. Returns false.
 */
bool err(char const *fmt, ...)
{
    va_list args;

    if (!fmt) {
	perror(thefilename);
	return false;
    }

    fprintf(stderr, "%s: ", thefilename ? thefilename : thisprog);
    va_start(args, fmt);
    vfprintf(stderr, fmt, args);
    fputc('\n', stderr);
    va_end(args);
    return false;
}

/* A wrapper around malloc() that dies if memory is unavailable.
 */
static void *allocate(size_t size)
{
    void *p = malloc(size);
    if (!p) {
        fputs("memory allocation failed\n", stderr);
        exit(EXIT_FAILURE);
    }
    return p;
}

/* A version of strdup() that dies if memory is unavailable.
 */
static char *strallocate(char const *str)
{
    size_t size;
    if (!str)
        return NULL;
    size = strlen(str) + 1;
    return memcpy(allocate(size), str, size);
}

/*
 * The functions that build the ELF file.
 */

/* Initialize connections between parts of the object file.
 */
static void connectparts(blueprint *bp)
{
    if (partispresent(&bp->parts[P_HASH]))
        sethashsyms(bp, P_HASH, P_DYNSYM);
    if (partispresent(&bp->parts[P_DYNSYM]))
        setsymstrings(bp, P_DYNSYM, P_DYNSTR);
    if (partispresent(&bp->parts[P_SYMTAB]))
        setsymstrings(bp, P_SYMTAB, P_STRTAB);
    if (partispresent(&bp->parts[P_REL])) {
        setrelsection(bp, P_REL, P_TEXT);
        setrelsyms(bp, P_REL, P_SYMTAB);
    }
    if (partispresent(&bp->parts[P_SHDRTAB]) &&
                partispresent(&bp->parts[P_SHSTRTAB]))
        setshdrstrs(bp, P_SHDRTAB, P_SHSTRTAB);
}

/* Set the names of the source file, the object file, and the
 * function. Names that were not set explicitly by the user are
 * derived from the name of the input file.
 */
static bool setnames(blueprint const *bp, int codetype, char const *filename)
{
    char const *base;
    int len, i;

    if (!filename || !filename[0])
        return false;
    base = strrchr(filename, '/');
    if (base)
        ++base;
    else
        base = filename;
    len = strlen(base);
    if (!len)
        return false;

    if (!srcfilename)
        srcfilename = strallocate(base);

    if (len > 2 && !memcmp(base + len - 2, ".b", 2))
	len -= 2;
    else if (len > 3 && !memcmp(base + len - 3, ".bf", 3))
        len -= 3;

    if (!outfilename) {
	if (bp->filetype == ET_REL) {
            outfilename = allocate(len + 3);
            sprintf(outfilename, "%*.*s.o", len, len, base);
	} else if (bp->filetype == ET_DYN) {
            outfilename = allocate(len + 7);
	    sprintf(outfilename, "lib%*.*s.so", len, len, base);
        } else if (bp->filetype == ET_EXEC && base[len] != '\0') {
            outfilename = allocate(len + 1);
            memcpy(outfilename, base, len);
            outfilename[len] = '\0';
        } else {
            outfilename = strallocate("a.out");
        }
    }

    if (!functionname) {
	if (codetype == CODE_EXEC) {
            functionname = strallocate("_start");
	} else {
            functionname = allocate(len + 1);
            memcpy(functionname, base, len);
            functionname[len] = '\0';
            if (!isalpha(functionname[0]))
                functionname[0] = '_';
	    for (i = 1 ; i < len ; ++i)
		if (!isalnum(functionname[i]))
		    functionname[i] = '_';
	}
    }

    return true;
}

/* Fill in the comment and text sections of the blueprint.
 */
static bool populateparts(blueprint *bp, int codetype,
                          char const *filename, int compressed)
{
    static char const comment[] = "\0ELF Brainfuck Compiler 2.0";

    if (addextras) {
        if (partispresent(&bp->parts[P_COMMENT])) {
	    bp->parts[P_COMMENT].shname = ".comment";
	    bp->parts[P_COMMENT].size = sizeof comment;
            bp->parts[P_COMMENT].part = allocate(sizeof comment);
            memcpy(bp->parts[P_COMMENT].part, comment, sizeof comment);
	}
        if (partispresent(&bp->parts[P_SYMTAB]))
	    addtosymtab(&bp->parts[P_SYMTAB], srcfilename,
			STB_LOCAL, STT_FILE, SHN_ABS);
    } else {
        if (bp->filetype == ET_EXEC) {
            removepart(&bp->parts[P_SHDRTAB]);
            removepart(&bp->parts[P_SHSTRTAB]);
        }
    }

    thefilename = filename;
    if (!compilebrainfuck(filename, bp, codetype, functionname, compressed))
	return false;
    thefilename = NULL;

    return true;
}

/* The top-level compiling function. Work through the stages of
 * creating the ELF file image, compiling the source code, and writing
 * out the file. The output file also has the executable bits in its
 * mode set accordingly.
 */
static bool compile(blueprint *bp, int codetype,
                    char const *filename, int compressed)
{
    struct stat	s;

    enforcevalidation(true);

    newparts(bp);

    connectparts(bp);
    initparts(bp);

    if (!setnames(bp, codetype, filename))
	return false;
    if (!populateparts(bp, codetype, filename, compressed))
	return false;

    fillparts(bp);
    measureparts(bp);

    if (!createfixups(bp, codetype, functionname))
        return false;

    completeparts(bp);

    thefilename = outfilename;
    if (!outputelf(bp, outfilename)) {
	err(NULL);
	remove(outfilename);
	return false;
    }
    if (stat(outfilename, &s) == 0) {
	if (bp->filetype == ET_EXEC)
	    s.st_mode |= S_IXUSR | S_IXGRP | S_IXOTH;
	else
	    s.st_mode &= ~(S_IXUSR | S_IXGRP | S_IXOTH);
	chmod(outfilename, s.st_mode);
    } else {
	err(NULL);
    }
    thefilename = NULL;

    return true;
}

/* main() uses the command-line options to set up the correct
 * blueprint, and then kicks off the compilation process.
 */
int main(int argc, char *argv[])
{
    static char const *optstring = "acf:i:lo:sxz";
    static struct option const options[] = {
        { "output", required_argument, NULL, 'o' },
        { "input", required_argument, NULL, 'i' },
        { "function", required_argument, NULL, 'f' },
        { "arg", no_argument, NULL, 'a' },
        { "strip", no_argument, NULL, 's' },
        { "help", no_argument, NULL, 'H' },
        { "version", no_argument, NULL, 'V' },
        { 0, 0, 0, 0 }
    };

    blueprint b;
    char const *filename;
    bool buildobjfile, bufferarg, compressed;
    int codetype;
    int ch, i;

    thisprog = argv[0];
    b.filetype = ET_NONE;
    codetype = CODE_NONE;
    buildobjfile = false;
    bufferarg = false;
    compressed = false;
    while ((ch = getopt_long(argc, argv, optstring, options, NULL)) != EOF) {
	switch (ch) {
	  case 'c':	buildobjfile = true;			break;
	  case 'x':	codetype = CODE_EXEC;			break;
	  case 'l':	codetype = CODE_SOFUNC;			break;
	  case 'z':	compressed = true;			break;
	  case 'a':	bufferarg = true;			break;
	  case 's':	addextras = false;			break;
	  case 'f':	functionname = strallocate(optarg);	break;
          case 'i':	srcfilename = strallocate(optarg);	break;
	  case 'o':	outfilename = strallocate(optarg);	break;
	  case 'H':	fputs(yowzitch, stdout);		return 0;
	  case 'V':	fputs(vourzhon, stdout);		return 0;
	  default:
            err("(try \"--help\" for more information)");
            exit(EXIT_FAILURE);
	}
    }
    if (optind + 1 != argc) {
	err(optind == argc ? "no source file specified."
			   : "multiple source files specified.");
	exit(EXIT_FAILURE);
    }
    filename = argv[optind];

    if (buildobjfile) {
        if (!codetype)
            codetype = CODE_FUNC;
        b.filetype = ET_REL;
    } else {
	if (!codetype)
	    codetype = CODE_EXEC;
	b.filetype = codetype == CODE_EXEC ? ET_EXEC : ET_DYN;
    }
    if (bufferarg) {
        if (codetype == CODE_EXEC) {
            err("executable file format cannot take an argument");
            exit(EXIT_FAILURE);
        }
        codetype = CODE_FUNCARG;
    }

    b.partcount = P_COUNT;
    b.parts = allocate(P_COUNT * sizeof *b.parts);
    memset(b.parts, 0, b.partcount * sizeof *b.parts);
    b.parts[0] = *parttable[P_EHDR];
    for (i = 0 ; partlists[b.filetype][i] ; ++i) {
        int type = partlists[b.filetype][i];
        b.parts[type] = *parttable[type];
    }
    if (!addextras)
        removepart(&b.parts[P_COMMENT]);

    return compile(&b, codetype, filename, compressed) ? EXIT_SUCCESS
						       : EXIT_FAILURE;
}
